compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala

```diff
@@ -665,8 +667,16 @@ trait ConstraintHandling {
     val wideInst =
       if isSingleton(bound) then inst
       else
-        val widenedFromSingle = widenSingle(inst)
-        val widenedFromUnion = widenOr(widenedFromSingle)
+        val widenedFromUnion =
+          if widenUnions && typeSize(inst) > 64 then
+            // If the inferred type `inst` is too large, the subtype check for `bound` in `widenSingle`
```

Comment:
Very useful diagnosis! I think we can even go further: If we have `widenUnions` then we never need to widen singletons in soft unions, since we will do an OrDominator afterwards, and this is the same for original or widen types. We don't even need a size limit for this.

I'll push a commit that tries this.

Reply:
I could fix it and account for transparent unions. See latest commit.

Reply:
Thanks, I updated the title.